cpp/references/data_structures/list_digraph.cpp

```diff
@@ -1 +1,596 @@
+}
+
+/* ########################### LIST DIGRAPH ########################### */
+
+
+/* ##### GDNATIVE API ##### */
+
+Array ListDigraph::add_vertices(Array p_data, bool p_return) {
+	Array ret;
+	for (int i = 0; i < p_data.size(); i++) {
+		LemonNode vertex(_graph.addNode());
+		(*_vertex_map)[vertex].data = p_data[i];
+
+		if (p_return) {
+			Ref<Vertex> instance(Vertex::create(vertex, this));
+			ret.append(instance);
+		}
+	}
+	return ret;
+}
+
+Ref<Arc> ListDigraph::add_arc(Ref<Vertex> p_source, Ref<Vertex> p_target) {
+	if (p_source.is_null()) {
+		ERR_NULL_PARAM(p_source);
+		return Ref<Arc>();
+	}
+	if (p_target.is_null()) {
+		ERR_NULL_PARAM(p_target);
+		return Ref<Arc>();
+	}
+	LemonArc arc(_graph.addArc(p_source->get_vertex(), p_target->get_vertex()));
+	Ref<Arc> instance(Arc::create(arc, this));
+	return instance;
+}
+
+Array ListDigraph::add_arcs(Array p_data, bool p_return) {
+	Array ret;
+	for (int i = 0; i < p_data.size(); i++) {
+		Variant vdict(p_data[i]);
+		if (vdict.get_type() != Variant::DICTIONARY)
+			continue;
+
+		Dictionary dict = vdict.operator Dictionary();
+		if (!dict.has("source") || !dict.has("target"))
+			continue;
+
+		Ref<Vertex> source(dict["source"]);
+		if (source.is_null())
+			continue;
+
+		Ref<Vertex> target(dict["target"]);
+		if (target.is_null())
+			continue;
+
+		LemonArc arc = _graph.addArc(source->get_vertex(), target->get_vertex());
+		(*_arc_map)[arc].data = (dict.has("data")) ? dict["data"] : Variant::NIL;
+
+		if (p_return) {
+			Ref<Arc> instance(Arc::create(arc, this));
+			ret.append(instance);
+		}
+	}
+	return ret;
+}
+
+void ListDigraph::erase_vertex(Ref<Vertex> p_vertex) {
+	if (p_vertex.is_null()) {
+		ERR_NULL_PARAM(p_vertex);
+		return;
+	}
+	_graph.erase(p_vertex->get_vertex());
+}
+
+void ListDigraph::erase_vertices(Array p_vertices) {
+	for (int i = 0; i < p_vertices.size(); i++) {
+		Ref<Vertex> vertex(p_vertices[i]);
+		if (vertex.is_null())
+			continue;
+		_graph.erase(vertex->get_vertex());
+	}
+}
+
+void ListDigraph::erase_arc(Ref<Arc> p_arc) {
+	if (p_arc.is_null()) {
+		ERR_NULL_PARAM(p_arc);
+		return;
+	}
+	_graph.erase(p_arc->get_arc());
+}
+
+void ListDigraph::erase_arcs(Array p_arcs) {
+	for (int i = 0; i < p_arcs.size(); i++) {
+		Ref<Arc> arc(p_arcs[i]);
+		if (arc.is_null())
+			continue;
+		_graph.erase(arc->get_arc());
+	}
+}
+
+void ListDigraph::contract(Ref<Vertex> p_vertex1, Ref<Vertex> p_vertex2, bool p_remove) {
+	if (p_vertex1.is_null()) {
+		ERR_NULL_PARAM(p_vertex1);
+		return;
+	}
+	if (p_vertex2.is_null()) {
+		ERR_NULL_PARAM(p_vertex2);
+		return;
+	}
+	_graph.contract(p_vertex1->get_vertex(), p_vertex2->get_vertex(), p_remove);
+}
+
+void ListDigraph::reserve_vertex(int p_amount) {
+	_graph.reserveNode(p_amount);
+}
+
+void ListDigraph::reserve_arc(int p_amount) {
+	_graph.reserveArc(p_amount);
+}
+
+Ref<Vertex> ListDigraph::get_vertex(int p_id) {
+	LemonNode vertex(_graph.nodeFromId(p_id));
+	if (!_graph.valid(vertex))
```

Comment:
We should make this an ERR_FAIL_COND_V replacement, i.e. `ERR_FAIL_COND_V(_graph.valid(vertex), NULL);` "If the condition is true, then report an error and return with the other parameter."